File: pi-tidy-tools/packages/pi-tidy-bots/sdk/python/tidy_backend_sdk/protocol.py
"""Wire validation. No gateway implementation or native runtime imports."""

import hashlib
import json
import math

DEFAULT_LIMITS = {
    "maxFrameBytes": 1024 * 1024,
    "maxUnacknowledgedEvents": 256,
    "maxSpoolBytes": 16 * 1024 * 1024,
    "maxPendingRequests": 256,
    "initializeTimeoutMs": 10000,
    "inspectTimeoutMs": 10000,
    "commandTimeoutMs": 15000,
    "shutdownTimeoutMs": 10000,
}
CORE_METHODS = ["health", "session.open", "session.snapshot", "operation.submit",
                "operation.inspect", "operation.cancel", "interaction.respond",
                "events.ack", "events.replay", "session.close", "shutdown"]
EVENT_TYPES = {"session.state", "operation.disposition", "turn.started", "message.started",
               "text.snapshot", "tool.started", "tool.updated", "tool.finished", "message.finished",
               "interaction.requested", "interaction.resolved", "usage.updated", "turn.terminal", "observation.gap"}
MUTATING = {"session.open", "operation.submit", "operation.cancel", "interaction.respond",
            "operation.steer", "session.compact", "session.configure", "session.import"}


class SDKError(Exception):
    def __init__(self, code, message=None):
        self.code = code
        super().__init__(message or code)


def integer(value, minimum=0):
    return type(value) is int and minimum <= value <= 9007199254740991


def identity(value):
    return isinstance(value, str) and bool(value) and len(value) <= 512 and "\0" not in value


def _json_value(value):
    if value is None or type(value) in (bool, str):
        return
    if type(value) is int and abs(value) <= 9007199254740991:
        return
    if type(value) is float and math.isfinite(value):
        return
    if isinstance(value, list):
        for entry in value:
            _json_value(entry)
        return
    if isinstance(value, dict) and all(isinstance(key, str) for key in value):
        for entry in value.values():
            _json_value(entry)
        return
    raise SDKError("invalid_payload", "Non-interoperable JSON value")


def canonical(value):
    _json_value(value)
    try:
        return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":"), allow_nan=False)
    except (ValueError, UnicodeError) as error:
        raise SDKError("invalid_payload") from error


def fingerprint(value):
    return "sha256:" + hashlib.sha256(canonical(value).encode("utf-8")).hexdigest()


def limits(value=None):
    result = dict(DEFAULT_LIMITS)
    for key, amount in (value or {}).items():
        if key not in result or not integer(amount, 1) or amount > DEFAULT_LIMITS[key]:
            raise SDKError("invalid_config", "Unsupported negotiated limit")
        result[key] = amount
    # Keep room for an explicit reliable gap, even when ordinary output is full.
    if result["maxUnacknowledgedEvents"] < 2 or result["maxFrameBytes"] < 4096 or result["maxSpoolBytes"] < 4096:
        raise SDKError("invalid_config", "Limits cannot reserve an observation-gap slot")
    return result


def parse_frame(raw):
    if not raw.endswith(b"\n"):
        raise SDKError("invalid_frame", "Unterminated frame")
    try:
        def unique(items):
            result = {}
            for key, value in items:
                if key in result:
                    raise ValueError("duplicate JSON key")
                result[key] = value
            return result
        value = json.loads(raw.decode("utf-8"), object_pairs_hook=unique,
                           parse_constant=lambda _: (_ for _ in ()).throw(ValueError("nonfinite")))
    except (ValueError, UnicodeError, RecursionError) as error:
        raise SDKError("invalid_frame") from error
    try:
        canonical(value).encode("utf-8")
        validate_rpc(value)
    except (SDKError, UnicodeError, RecursionError) as error:
        raise SDKError("invalid_frame") from error
    return value


def validate_rpc(value):
    if not isinstance(value, dict) or value.get("jsonrpc") != "2.0":
        raise SDKError("invalid_frame")
    if "id" in value and not isinstance(value["id"], str):
        raise SDKError("invalid_frame", "RPC IDs are strings")
    if "method" in value:
        if not identity(value["method"]) or "result" in value or "error" in value or ("params" in value and not isinstance(value["params"], dict)):
            raise SDKError("invalid_frame")
    elif "id" not in value or "params" in value or ("result" in value) == ("error" in value):
        raise SDKError("invalid_frame")
    elif "error" in value:
        error = value["error"]
        if not isinstance(error, dict) or type(error.get("code")) is not int or not isinstance(error.get("message"), str):
            raise SDKError("invalid_frame")


def encode_frame(value, maximum):
    validate_rpc(value)
    try:
        frame = (canonical(value) + "\n").encode("utf-8")
    except (UnicodeError, RecursionError) as error:
        raise SDKError("invalid_frame") from error
    if len(frame) > maximum:
        raise SDKError("resource_limit", "Complete encoded frame exceeds its byte budget")
    return frame


def validate_event(event):
    if not isinstance(event, dict) or event.get("type") not in EVENT_TYPES or not isinstance(event.get("payload"), dict):
        raise SDKError("invalid_event")
    if not identity(event.get("eventId")) or not identity(event.get("bindingId")) or not integer(event.get("leaseGeneration"), 1) or not integer(event.get("sourceSequence"), 1):
        raise SDKError("invalid_event")
    for name in ("operationId", "turnId", "messageId", "blockId", "toolCallId", "interactionId"):
        if name in event and not identity(event[name]):
            raise SDKError("invalid_event")
    kind = event["type"]
    if kind not in ("session.state", "observation.gap") and not all(identity(event.get(name)) for name in ("operationId", "turnId")):
        raise SDKError("invalid_event", "Missing turn correlation")
    if (kind.startswith("message.") or kind == "text.snapshot") and not identity(event.get("messageId")):
        raise SDKError("invalid_event")
    if kind == "text.snapshot" and (not identity(event.get("blockId")) or not integer(event["payload"].get("revision")) or not isinstance(event["payload"].get("text"), str)):
        raise SDKError("invalid_event")
    if kind.startswith("tool.") and not identity(event.get("toolCallId")):
        raise SDKError("invalid_event")
    if kind.startswith("interaction.") and not identity(event.get("interactionId")):
        raise SDKError("invalid_event")
    payload = event["payload"]
    if kind == "operation.disposition" and payload.get("disposition") not in ("accepted", "rejected", "unknown"):
        raise SDKError("invalid_event")
    if kind == "message.started" and (payload.get("role") not in ("assistant", "user", "tool", "system") or not integer(payload.get("order"))):
        raise SDKError("invalid_event")
    if kind == "message.finished":
        if not isinstance(payload.get("blocks"), list):
            raise SDKError("invalid_event")
        for block in payload["blocks"]:
            if not isinstance(block, dict) or not identity(block.get("blockId")) or not integer(block.get("revision")):
                raise SDKError("invalid_event")
            if block.get("type") == "text" and isinstance(block.get("text"), str):
                continue
            if block.get("type") == "artifact" and identity(block.get("artifactId")) and identity(block.get("mediaType")):
                continue
            raise SDKError("invalid_event")
    if kind == "turn.terminal" and (payload.get("execution") not in ("ended", "failed", "cancelled", "interrupted") or payload.get("observation") not in ("complete", "live_gap", "reconciliation_required")):
        raise SDKError("invalid_event")
    if "contextBudget" in payload:
        budget = payload["contextBudget"]
        if (
            not isinstance(budget, dict)
            or not integer(budget.get("remainingTokens"))
            or budget.get("source") not in ("adapter", "gateway")
            or ("usedTokens" in budget and not integer(budget.get("usedTokens")))
            or ("windowTokens" in budget and not integer(budget.get("windowTokens"), 1))
        ):
            raise SDKError("invalid_event")


def validate_capabilities(value):
    fields = {
        "input": {"text", "mediaTypes", "maxMediaBytes"},
        "sessions": {"load", "import", "continuity"},
        "output": {"text", "tools", "usage"},
        "operations": {"nativeDedupe", "nativeDedupeRetentionMs", "nativeReplay", "nativeReplayRetentionMs", "nativeReplayGapSemantics", "cancel", "steer"},
        "interactions": {"permissions", "questions"},
        "configuration": {"model", "thinking", "compact", "new_context"},
    }
    if not isinstance(value, dict):
        raise SDKError("invalid_capabilities")
    for section, names in fields.items():
        if not isinstance(value.get(section), dict):
            raise SDKError("invalid_capabilities")
        for key, child in value[section].items():
            if key not in names and ("." not in key or isinstance(child, dict) and child.get("required") is True):
                raise SDKError("invalid_capabilities")
    for key, child in value.items():
        if key not in fields and key != "fleetTools" and ("." not in key or isinstance(child, dict) and child.get("required") is True):
            raise SDKError("invalid_capabilities")
    inp = value["input"]
    if inp.get("text") is not True or not isinstance(inp.get("mediaTypes"), list) or not all(identity(v) for v in inp["mediaTypes"]) or not integer(inp.get("maxMediaBytes")):
        raise SDKError("invalid_capabilities")
    for section, names in {"sessions": ["load", "import"], "output": ["tools"], "operations": ["steer"], "interactions": ["questions"], "configuration": ["model", "thinking", "compact"]}.items():
        if any(type(value[section].get(name)) is not bool for name in names):
            raise SDKError("invalid_capabilities")
    if "new_context" in value["configuration"] and type(value["configuration"].get("new_context")) is not bool:
        raise SDKError("invalid_capabilities")
    if type(value.get("fleetTools")) is not bool:
        raise SDKError("invalid_capabilities")
    for section, name, allowed in [
        ("sessions", "continuity", ["verified", "unverified"]), ("output", "text", ["final-only", "snapshots"]),
        ("output", "usage", ["reported", "estimated", "unknown"]), ("operations", "nativeDedupe", ["durable", "none"]),
        ("operations", "nativeReplay", ["cursor", "none"]), ("operations", "cancel", ["cooperative", "process-termination", "unsupported"]),
        ("interactions", "permissions", ["exact-request", "none"]),
    ]:
        if value[section].get(name) not in allowed:
            raise SDKError("invalid_capabilities")
    ops = value["operations"]
    if ops["nativeDedupe"] == "durable" and not integer(ops.get("nativeDedupeRetentionMs"), 1):
        raise SDKError("invalid_capabilities")
    if ops["nativeReplay"] == "cursor" and (not integer(ops.get("nativeReplayRetentionMs"), 1) or ops.get("nativeReplayGapSemantics") != "explicit-gap"):
        raise SDKError("invalid_capabilities")
    if (value["sessions"]["load"] or value["sessions"]["import"]) and value["sessions"]["continuity"] != "verified":
        raise SDKError("invalid_capabilities")
    return value

File: pi-tidy-tools/packages/pi-tidy-bots/src/gateway/protocol.ts
import { TextDecoder } from "node:util";

export const DEFAULT_LIMITS = Object.freeze({
  maxFrameBytes: 1024 * 1024,
  maxUnacknowledgedEvents: 256,
  maxSpoolBytes: 16 * 1024 * 1024,
  maxPendingRequests: 256,
  initializeTimeoutMs: 10_000,
  inspectTimeoutMs: 10_000,
  commandTimeoutMs: 15_000,
  shutdownTimeoutMs: 10_000,
});
export type ProtocolLimits = { [K in keyof typeof DEFAULT_LIMITS]: number };
export type JsonObject = Record<string, unknown>;
export type RpcMessage = {
  jsonrpc: "2.0";
  id?: string;
  method?: string;
  params?: JsonObject;
  result?: unknown;
  error?: { code: number; message: string; data?: unknown };
};
export class ProtocolError extends Error {
  readonly code: string;
  constructor(code: string, message: string) {
    super(message);
    this.code = code;
    this.name = "ProtocolError";
  }
}
export function object(value: unknown): value is JsonObject {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
export function nonempty(value: unknown): value is string {
  return typeof value === "string" && value.length > 0;
}
export function validateLimits(
  limits: Partial<ProtocolLimits> = {}
): ProtocolLimits {
  const output = { ...DEFAULT_LIMITS, ...limits };
  for (const [key, value] of Object.entries(output)) {
    if (
      !(key in DEFAULT_LIMITS) ||
      !Number.isSafeInteger(value) ||
      value <= 0 ||
      value > DEFAULT_LIMITS[key as keyof ProtocolLimits]
    )
      throw new ProtocolError(
        "invalid_config",
        `Invalid negotiated limit ${key}`
      );
  }
  return output;
}
export function parseRpc(value: unknown): RpcMessage {
  if (!object(value) || value.jsonrpc !== "2.0")
    throw new ProtocolError(
      "invalid_frame",
      "Expected a JSON-RPC 2.0 object; batches are unsupported"
    );
  const has = (key: string) => Object.hasOwn(value, key);
  if (has("id") && typeof value.id !== "string")
    throw new ProtocolError("invalid_frame", "RPC IDs must be strings");
  if (has("method")) {
    if (
      !nonempty(value.method) ||
      has("result") ||
      has("error") ||
      (has("params") && !object(value.params))
    )
      throw new ProtocolError("invalid_frame", "Invalid RPC request");
  } else {
    if (!has("id") || has("params") || has("result") === has("error"))
      throw new ProtocolError("invalid_frame", "Invalid RPC response");
    if (
      has("error") &&
      (!object(value.error) ||
        !Number.isInteger(value.error.code) ||
        typeof value.error.message !== "string")
    )
      throw new ProtocolError("invalid_frame", "Invalid RPC error");
  }
  return value as RpcMessage;
}
/** Counts the complete encoded envelope, including its LF, before any pipe write. */
export function encodeFrame(
  message: RpcMessage,
  maxBytes: number = DEFAULT_LIMITS.maxFrameBytes
): Buffer {
  parseRpc(message);
  const frame = Buffer.from(JSON.stringify(message) + "\n", "utf8");
  if (frame.length > maxBytes)
    throw new ProtocolError(
      "resource_limit",
      "Complete protocol frame exceeds negotiated byte limit"
    );
  return frame;
}
/** A bounded byte parser: LF alone terminates a frame, never Unicode separators. */
export class FrameDecoder {
  readonly maxBytes: number;
  private readonly buffer: Buffer;
  private size = 0;
  private readonly decoder = new TextDecoder("utf-8", { fatal: true });
  constructor(maxBytes: number = DEFAULT_LIMITS.maxFrameBytes) {
    this.maxBytes = maxBytes;
    if (
      !Number.isSafeInteger(maxBytes) ||
      maxBytes < 1 ||
      maxBytes > DEFAULT_LIMITS.maxFrameBytes
    )
      throw new ProtocolError("invalid_config", "Invalid frame byte limit");
    this.buffer = Buffer.allocUnsafe(maxBytes - 1);
  }
  push(chunk: Buffer, receive: (message: RpcMessage) => void): void {
    let offset = 0;
    while (offset < chunk.length) {
      const end = chunk.indexOf(10, offset);
      const stop = end === -1 ? chunk.length : end;
      const slice = chunk.subarray(offset, stop);
      // Reserve the mandatory LF even while it has not arrived.
      if (this.size + slice.length + 1 > this.maxBytes)
        throw new ProtocolError(
          "resource_limit",
          "Plugin frame exceeds negotiated byte limit"
        );
      if (slice.length) {
        slice.copy(this.buffer, this.size);
        this.size += slice.length;
      }
      if (end === -1) return;
      const bytes = this.buffer.subarray(0, this.size);
      this.size = 0;
      let value: unknown;
      try {
        value = JSON.parse(this.decoder.decode(bytes));
      } catch {
        throw new ProtocolError(
          "invalid_frame",
          "Plugin emitted malformed JSON or UTF-8"
        );
      }
      receive(parseRpc(value));
      offset = end + 1;
    }
  }
  finish(): void {
    if (this.size)
      throw new ProtocolError(
        "invalid_frame",
        "Plugin exited with an unterminated frame"
      );
  }
}

export type SessionProof = "none" | "identity-only" | "retained-history";
export type EmptySeatPolicy = "non-restorable" | "restartable";
export type ContinuityStatus = "verified" | "unverified";
export type SessionEvidenceProvenance =
  | "native-identity"
  | "codex-thread-identity"
  | "pi-history-checkpoint"
  | "hermes-checkpoint-v1";

/** Load support, proof type, and evidence provenance are distinct.
 * `load`/`import` mean restore is attempted. `proof` is what a successful
 * load actually proved. `continuity: verified` means that advertised proof
 * succeeded — not that every backend has Pi/Hermes-equivalent history.
 * Never-prompted seats may stay `emptySeat: non-restorable`; fail closed
 * rather than invent restart-availability.
 */
export interface CapabilityDescriptor {
  input: { text: true; mediaTypes: string[]; maxMediaBytes: number };
  sessions: {
    load: boolean;
    import: boolean;
    continuity: ContinuityStatus;
    proof?: SessionProof;
    emptySeat?: EmptySeatPolicy;
  };
  output: {
    text: "final-only" | "snapshots";
    tools: boolean;
    usage: "reported" | "estimated" | "unknown";
  };
  operations: {
    nativeDedupe: "durable" | "none";
    nativeDedupeRetentionMs?: number;
    nativeReplay: "cursor" | "none";
    nativeReplayRetentionMs?: number;
    nativeReplayGapSemantics?: "explicit-gap";
    cancel: "cooperative" | "process-termination" | "unsupported";
    steer: boolean;
  };
  interactions: { permissions: "exact-request" | "none"; questions: boolean };
  configuration: {
    model: boolean;
    thinking: boolean;
    compact: boolean;
    new_context?: boolean;
  };
  fleetTools: boolean;
  [extension: string]: unknown;
}
export const CORE_METHODS = [
  "health",
  "session.open",
  "session.snapshot",
  "operation.submit",
  "operation.inspect",
  "operation.cancel",
  "interaction.respond",
  "events.ack",
  "events.replay",
  "session.close",
  "shutdown",
] as const;
export const EVENT_TYPES = new Set([
  "session.state",
  "operation.disposition",
  "turn.started",
  "message.started",
  "text.snapshot",
  "tool.started",
  "tool.updated",
  "tool.finished",
  "message.finished",
  "interaction.requested",
  "interaction.resolved",
  "usage.updated",
  "turn.terminal",
  "observation.gap",
]);
export interface GatewayPluginEvent extends JsonObject {
  bindingId: string;
  leaseGeneration: number;
  sourceSequence: number;
  eventId: string;
  type: string;
  payload: JsonObject;
  operationId?: string;
  turnId?: string;
}
export function validateEvent(value: unknown): GatewayPluginEvent {
  if (
    !object(value) ||
    !nonempty(value.bindingId) ||
    !Number.isSafeInteger(value.leaseGeneration) ||
    Number(value.leaseGeneration) < 1 ||
    !Number.isSafeInteger(value.sourceSequence) ||
    Number(value.sourceSequence) < 1 ||
    !nonempty(value.eventId) ||
    typeof value.type !== "string" ||
    !EVENT_TYPES.has(value.type) ||
    !object(value.payload)
  )
    throw new ProtocolError(
      "invalid_event",
      "Malformed canonical plugin event"
    );
  for (const key of [
    "operationId",
    "turnId",
    "messageId",
    "blockId",
    "toolCallId",
    "interactionId",
  ])
    if (Object.hasOwn(value, key) && !nonempty(value[key]))
      throw new ProtocolError("invalid_event", `Invalid ${key}`);
  if (
    value.type !== "session.state" &&
    value.type !== "observation.gap" &&
    (!nonempty(value.operationId) || !nonempty(value.turnId))
  )
    throw new ProtocolError(
      "invalid_event",
      "Turn event requires operation and turn identity"
    );
  if (
    (value.type.startsWith("message.") || value.type === "text.snapshot") &&
    !nonempty(value.messageId)
  )
    throw new ProtocolError(
      "invalid_event",
      "Message event requires message identity"
    );
  if (
    value.type === "text.snapshot" &&
    (!nonempty(value.blockId) ||
      !Number.isSafeInteger(value.payload.revision) ||
      Number(value.payload.revision) < 0 ||
      typeof value.payload.text !== "string")
  )
    throw new ProtocolError(
      "invalid_event",
      "Snapshot requires block, revision and text"
    );
  if (value.type.startsWith("tool.") && !nonempty(value.toolCallId))
    throw new ProtocolError(
      "invalid_event",
      "Tool event requires tool identity"
    );
  if (value.type.startsWith("interaction.") && !nonempty(value.interactionId))
    throw new ProtocolError(
      "invalid_event",
      "Interaction event requires exact request identity"
    );
  const payload = value.payload;
  if (
    value.type === "operation.disposition" &&
    !["accepted", "rejected", "unknown"].includes(String(payload.disposition))
  )
    throw new ProtocolError(
      "invalid_event",
      "Invalid native delivery disposition"
    );
  if (
    value.type === "message.started" &&
    (!["assistant", "user", "tool", "system"].includes(String(payload.role)) ||
      !Number.isSafeInteger(payload.order) ||
      Number(payload.order) < 0)
  )
    throw new ProtocolError(
      "invalid_event",
      "Message start requires role and stable order"
    );
  if (value.type === "message.finished") {
    if (!Array.isArray(payload.blocks))
      throw new ProtocolError(
        "invalid_event",
        "Final message requires authoritative blocks"
      );
    for (const block of payload.blocks) {
      if (
        !object(block) ||
        !nonempty(block.blockId) ||
        !Number.isSafeInteger(block.revision) ||
        Number(block.revision) < 0 ||
        (block.type === "text"
          ? typeof block.text !== "string"
          : block.type === "artifact"
            ? !nonempty(block.artifactId) || !nonempty(block.mediaType)
            : true)
      )
        throw new ProtocolError(
          "invalid_event",
          "Malformed canonical final block"
        );
    }
  }
  if (
    value.type === "turn.terminal" &&
    (!["ended", "failed", "cancelled", "interrupted"].includes(
      String(payload.execution)
    ) ||
      !["complete", "live_gap", "reconciliation_required"].includes(
        String(payload.observation)
      ))
  )
    throw new ProtocolError(
      "invalid_event",
      "Terminal event requires explicit execution and observation state"
    );
  if (Object.hasOwn(payload, "contextBudget"))
    validateContextBudget(payload.contextBudget);
  return value as GatewayPluginEvent;
}

function validateContextBudget(value: unknown): void {
  if (
    !object(value) ||
    !Number.isSafeInteger(value.remainingTokens) ||
    Number(value.remainingTokens) < 0 ||
    !["adapter", "gateway"].includes(String(value.source)) ||
    (Object.hasOwn(value, "usedTokens") &&
      (!Number.isSafeInteger(value.usedTokens) ||
        Number(value.usedTokens) < 0)) ||
    (Object.hasOwn(value, "windowTokens") &&
      (!Number.isSafeInteger(value.windowTokens) ||
        Number(value.windowTokens) < 1))
  )
    throw new ProtocolError("invalid_event", "Malformed contextBudget");
}
export function validateCapabilities(value: unknown): CapabilityDescriptor {
  const fail = (message: string): never => {
    throw new ProtocolError("invalid_capabilities", message);
  };
  if (!object(value)) return fail("Missing capabilities");
  const sections: Record<string, string[]> = {
    input: ["text", "mediaTypes", "maxMediaBytes"],
    sessions: ["load", "import", "continuity", "proof", "emptySeat"],
    output: ["text", "tools", "usage"],
    operations: [
      "nativeDedupe",
      "nativeDedupeRetentionMs",
      "nativeReplay",
      "nativeReplayRetentionMs",
      "nativeReplayGapSemantics",
      "cancel",
      "steer",
    ],
    interactions: ["permissions", "questions"],
    configuration: ["model", "thinking", "compact", "new_context"],
  };
  for (const [name, keys] of Object.entries(sections)) {
    const section = value[name];
    if (!object(section)) return fail(`Missing ${name} capabilities`);
    for (const key of Object.keys(section)) {
      if (
        !keys.includes(key) &&
        (!key.includes(".") ||
          (object(section[key]) && section[key].required === true))
      )
        return fail(`Unknown required capability ${name}.${key}`);
    }
  }
  for (const key of Object.keys(value))
    if (
      !(key in sections) &&
      key !== "fleetTools" &&
      (!key.includes(".") ||
        (object(value[key]) && value[key].required === true))
    )
      return fail(`Unknown required capability ${key}`);
  const descriptor = value as unknown as CapabilityDescriptor;
  if (
    descriptor.input.text !== true ||
    !Array.isArray(descriptor.input.mediaTypes) ||
    !descriptor.input.mediaTypes.every(nonempty) ||
    !Number.isSafeInteger(descriptor.input.maxMediaBytes) ||
    descriptor.input.maxMediaBytes < 0
  )
    return fail("Invalid text/media input contract");
  const booleans = [
    descriptor.sessions.load,
    descriptor.sessions.import,
    descriptor.output.tools,
    descriptor.operations.steer,
    descriptor.interactions.questions,
    descriptor.configuration.model,
    descriptor.configuration.thinking,
    descriptor.configuration.compact,
    descriptor.fleetTools,
  ];
  if (!booleans.every((entry) => typeof entry === "boolean"))
    return fail("Missing boolean capability");
  if (
    descriptor.configuration.new_context !== undefined &&
    typeof descriptor.configuration.new_context !== "boolean"
  )
    return fail("Invalid new_context capability");
  const enums: [unknown, string[]][] = [
    [descriptor.sessions.continuity, ["verified", "unverified"]],
    [descriptor.output.text, ["final-only", "snapshots"]],
    [descriptor.output.usage, ["reported", "estimated", "unknown"]],
    [descriptor.operations.nativeDedupe, ["durable", "none"]],
    [descriptor.operations.nativeReplay, ["cursor", "none"]],
    [
      descriptor.operations.cancel,
      ["cooperative", "process-termination", "unsupported"],
    ],
    [descriptor.interactions.permissions, ["exact-request", "none"]],
  ];
  if (enums.some(([entry, allowed]) => !allowed.includes(String(entry))))
    return fail("Unknown capability value");
  const restore = descriptor.sessions.load || descriptor.sessions.import;
  const proof = descriptor.sessions.proof ?? "none";
  if (!["none", "identity-only", "retained-history"].includes(proof))
    return fail("Unknown session proof");
  if (
    descriptor.sessions.emptySeat !== undefined &&
    descriptor.sessions.emptySeat !== "non-restorable" &&
    descriptor.sessions.emptySeat !== "restartable"
  )
    return fail("Unknown empty-seat policy");
  if (restore) {
    if (descriptor.sessions.continuity !== "verified")
      return fail("Session restoration requires verified continuity");
    if (proof === "none")
      return fail("Session restoration requires an explicit proof level");
    if (descriptor.sessions.emptySeat === undefined)
      return fail("Session restoration requires an explicit empty-seat policy");
  } else if (proof !== "none")
    return fail("Proof without load or import overclaims restoration");
  const positive = (entry: unknown) =>
    Number.isSafeInteger(entry) && Number(entry) > 0;
  if (
    descriptor.operations.nativeDedupe === "durable" &&
    !positive(descriptor.operations.nativeDedupeRetentionMs)
  )
    return fail("Durable dedupe requires a retention window");
  if (
    descriptor.operations.nativeReplay === "cursor" &&
    (!positive(descriptor.operations.nativeReplayRetentionMs) ||
      descriptor.operations.nativeReplayGapSemantics !== "explicit-gap")
  )
    return fail("Cursor replay requires retention and explicit gap semantics");
  return descriptor;
}

export function sessionProofOf(
  sessions: CapabilityDescriptor["sessions"]
): SessionProof {
  return sessions.proof ?? "none";
}

export function emptySeatOf(
  sessions: CapabilityDescriptor["sessions"]
): EmptySeatPolicy {
  return sessions.emptySeat ?? "non-restorable";
}

export function sessionOpenEvidenceMatches(
  result: JsonObject,
  advertised: SessionProof
): boolean {
  if (advertised === "none")
    return result.proof === "none" || result.proof === undefined;
  if (result.continuity !== "verified" || result.proof !== advertised)
    return false;
  if (!object(result.evidence) || !nonempty(result.evidence.provenance))
    return false;
  const provenance = String(result.evidence.provenance);
  if (advertised === "identity-only")
    return (
      provenance !== "pi-history-checkpoint" &&
      provenance !== "hermes-checkpoint-v1"
    );
  return (
    provenance === "pi-history-checkpoint" ||
    provenance === "hermes-checkpoint-v1"
  );
}
